source/extensions/common/wasm/plugin.cc

```diff
@@ -1,25 +1,80 @@
 #include "extensions/common/wasm/plugin.h"
 
+#include <cstdlib>
 #include <memory>
 
+#include "envoy/common/exception.h"
 #include "envoy/extensions/wasm/v3/wasm.pb.validate.h"
 #include "envoy/local_info/local_info.h"
 
 #include "common/protobuf/protobuf.h"
 #include "common/protobuf/utility.h"
 
+#include "extensions/common/wasm//well_known_names.h"
+
 #include "include/proxy-wasm/wasm.h"
 
 namespace Envoy {
 namespace Extensions {
 namespace Common {
 namespace Wasm {
 
+void setEnvVar(const std::string& name, const std::string& value, int overwrite) {
+#ifdef WIN32
+  if (!overwrite) {
+    size_t requiredSize;
+    ::getenv_s(&requiredSize, nullptr, 0, name.c_str());
+    if (requiredSize != 0) {
+      return;
+    }
+  }
+  ::_putenv_s(name.c_str(), value.c_str());
+#else
+  ::setenv(name.c_str(), value.c_str(), overwrite);
+#endif
+}
+
 WasmConfig::WasmConfig(const envoy::extensions::wasm::v3::PluginConfig& config) : config_(config) {
   for (auto& capability : config_.capability_restriction_config().allowed_capabilities()) {
     // TODO(rapilado): Set the SanitizationConfig fields once sanitization is implemented.
     allowed_capabilities_[capability.first] = proxy_wasm::SanitizationConfig();
   }
+
+  auto has_env_config = config_.vm_config().has_environment_variables();
+  if (has_env_config && config.vm_config().runtime() == WasmRuntimeNames::get().Null) {
+    // On NullVm, std::getenv directly accesses to the Envoy's env vars, so we need to set here to
+    // be consistent with actual Wasm VMs.
+    for (auto& env : config_.vm_config().environment_variables().key_values()) {
+      setEnvVar(env.first, env.second, true);
+    }
+  } else if (has_env_config) {
+#define DUPLICATED_ENV_KEY_EXCEPTION(key, name)                                                    \
+  throw EnvoyException(                                                                            \
+      fmt::format("Key {} is duplicated in "                                                       \
+                  "envoy.extensions.wasm.v3.VmConfig.environment_variables for {}. "               \
+                  "All the keys must be unique.",                                                  \
+                  key, name))
+    auto envs = config_.vm_config().environment_variables();
+    for (auto& key : envs.host_env_keys()) {
+      if (envs_.find(key) == envs_.end()) {
+        auto value = std::getenv(key.data());
+        if (value) {
```

Comment:
This is interesting case, because if someone configures:
```
host_env_keys: ["FOO"]
key_values: [{"FOO": "BAR"}]
```
then it will throw an exception if `FOO` exists in Envoy's environment, but it will be accepted if it doesn't.

It think the configuration should be rejected regardless, i.e. you need to compare keys in configuration.